service/resolver.go

```diff
@@ -0,0 +1,73 @@
+package service
+
+import (
+	"fmt"
+	"math/rand"
+	"net/url"
+	"time"
+
+	"sync"
+	"sync/atomic"
+	"unsafe"
+
+	"github.com/hashicorp/consul/api"
+)
+
+// Resolver for discovery service
+type Resolver struct {
+	consulClient    *api.Client
+	endpoints       []*url.URL
+	currentEndpoint *url.URL
+	generator       *rand.Rand
+	TTL             int64
+	lock            sync.Mutex
+}
+
+// GetNodesFromConsul get service nodes form service discovery
+func (r *Resolver) GetNodesFromConsul(service string) (entries []*api.ServiceEntry) {
+	entries = make([]*api.ServiceEntry, 0)
+	nodes, _, error := r.consulClient.Health().Service(service, "", true, &api.QueryOptions{
+		AllowStale:        true,
+		RequireConsistent: false,
+	})
+	if error == nil {
+		entries = nodes
+	}
+	return
+}
+
+func (r *Resolver) prepareCurrentEndpoint() (currentEndpoint *url.URL) {
```

Comment:
What `current` means in this context? Maybe something like `getHealthyInstanceEndpoint`